keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,382 @@
+### User Stories (Optional)
+
+#### Story 1
+
+Kubernetes by default sends a SIGTERM to all containers while killing them. When running nginx on Kubernetes, this can result in nginx dropping requests as reported [here](https://github.com/Kong/kubernetes-ingress-controller/pull/283). The current solution for this issue would be to build custom images with a SIGQUIT stop signal or to write a PreStop lifecycle hook that manually kills the process gracefully, which is what is done in the PR. If we had stop signal support at the Container spec level, this would've been easier and straightforward to implement. Users wouldn't have to patch the applications running on Kubernetes to handle different termination behavior. This is also similar to [this issue](https://github.com/github/resque/pull/21). 
+
+### Risks and Mitigations
+
+I don't see any issues with adding a new field for the stop signal because this is an optional feature that users.
+
+## Design Details
+
+On top of the details described in the [Proposal](#proposal), these are some details on how exactly the new field will work.
+- `StopSignal` in the Container spec can be a nil value. In this case, the default stop signal of SIGTERM will be used.
+- The `StopSignal` field will override the stop signal set from the container image definition.
+- The order of priority for the different stop signals would look like this
+	`StopSignal from Container Spec > STOPSIGNAL from container image > Default value of SIGTERM`
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Alpha:
+- Test that the validation fails when given a non string value for StopSignal
+- Test that the validation passes when given a proper string value representing a standard stop signal
+- Test that the validation fails when the StopSignal is given with the feature gate disabled
+- Test that the validation returns the appropriate error message when an invalid string value is given for StopSignal
+- Tests for verifying behavior when feature gate is disabled after being used to create Pods where StopSignal field is used
+- Tests for verifying behavior when feature gate is reenabled after being disabled after creating Pods with StopSignal
+
+##### Integration tests
+  
+##### e2e tests
+
+- Test that containers are killed with the right stop signal when StopSignal is passed
+- Test that containers are killed with SIGTERM when no StopSignal is passed
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- CRI API implementation completed in containerd marked as experimental
+- Initial e2e tests completed and enabled, testing the feature against containerd
+- Unit tests for validation, e2e tests for version skew
+
+#### Beta
+
+- CRI API implementation for CRI-O
```

Comment:
@haircommander I think this should be simple enough to get alpha support in CRI-O as well.

Reply:
@mrunalp I can work on CRI-O in alpha since the change should be small-ish but what do you think about adding Windows support in beta? Do we usually ship Windows support in alpha itself? Especially for CRI related KEPs.

Reply:
First we need to define what Windows support means.  Does Windows have a sigterm equivalent when container is stopped?

Reply:
AFAIK windows is usually best effort. Many examples I see windows is added as a separate KEP if the windows community is interested (CPU Manager, graceful node shutdown). I'm not sure if this is good or bad though.

I think windows support in beta could be good but I don't know if it needs to block the current KEP. That would be more of a question for sig-windows.

Reply:
We should target CRI-O for alpha with the new container runtime policy. We could bring this to beta gate off with just containerd but we would need CRI-O to turn it on for beta. So I think adding this in alpha should be fine.

Reply:
My point with Windows is we should not just say "we'll maybe handle windows later" without even knowing what Windows does.

When a docker container on Windows is stopped, what is the Windows equivalent of STOPSIGNAL?  https://github.com/moby/moby/issues/25982

If we add a field which only works on Linux, are we making a problem for future us?

Reply:
we'll test it on windows as well

Reply:
Pushed another commit adding more context about hcsshim and how it supports SIGTERM and SIGKILL as valid signals: https://github.com/kubernetes/enhancements/pull/5122/commits/d29143f957d02b061ac6a0db7d26a2e71cde1cd2